### ISSUE TITLE
Setting 'login' in config['skin_logo'] changes also the favicon accordingly

### ISSUE THREAD

**Issue by gianlucagiacometti** [roundcube/roundcubemail#7456]
To reproduce the issue (only in the login page):

```
$config['skin_logo'] = array(
  'login' = '[a valid image]'
);
```

sets the logo in the login page but also the favicon with the same image

or with

```
$config['skin_logo'] = array(
  'login' = 'whatever which is not a valid image, also this text, but not an empty string'
);

```
No favicon displayed, the alt text 'Logo' shows.

No way to visualise the correct favicon when login is set in the array.

Last version in master branch used (downloaded today)

**Comment by johndoh** [comment]
> No way to visualise the correct favicon when login is set in the array.

Set the logo for `favicon` type e.g.:

```
'[favicon]' => '/my/favicon.png'
```

As for matching the logo template this may be a regression caused by 75125cbac16b6ad714e3bd019f218e1d1794f9e6 which removed the strict matching param. I'm not certain of the best approach. This line:
https://github.com/roundcube/roundcubemail/blob/98881a0cbd74936240d7f9afcba7223d60d64241/program/include/rcmail_output_html.php#L2517-L2520
Could be changed to:
```php
$template_names = preg_grep("/\]$/", $template_names);
```
so that only entries matching the specified type are used. That makes sense for the `[favicon]` and I think also for `[small]` the only one I'm not certain of is `[print]` if I specify a logo for a skin then I might expect the print logo to also change without additional config.

**Comment by alecpl** [comment]
I'm also not sure about this.

BTW, @johndoh,  isn't the regexp wrong in `$template_names = preg_grep("/\*$/", $template_names, PREG_GREP_INVERT);`? There's only one name that ends with a star. Also, the other regexp, has inconsistent use of backslash.

**Comment by johndoh** [comment]
I think when I did 75125cb I got that line wrong and that it does not make sense any more. Like you said there is only 2 entries which end in a * and thinking about it today removing only those when a logo type is specified seems arbitrary and kind of pointless.

I think that line could be removed completely or replaced with something that limits the matches to those that specify the type. Both approaches have advantages and disadvantages.

May be I need to bring back the strict matching parameter or perhaps allow the logo object to specify multiple types so it define a fallback.

**Comment by johndoh** [comment]
This needs more testing but I might do the trick. With this change the `src` attribute is checked to decide if a logo is "expected" or "optional". This means that for types like `favicon` or `small` nothing will be returned unless it is explicit defined in the config - prevents it from mixing up the logo to show on the login screen and the image to use for the favicon. It also handles the `print` case where in Elastic there is no logo by default (so `src` is empty) and so no logo is returned unless explicitly defined in config using `[print] => '/my/logo.png'` and for other skins which have a print logo by default then it checks for config like `'[print]' => '/ny/logo.png'` but also for ones like `'larry:*' => '/my/logo.png'`.

```diff
diff --git a/program/include/rcmail_output_html.php b/program/include/rcmail_output_html.php
index ceb80726a..7e4750af8 100644
--- a/program/include/rcmail_output_html.php
+++ b/program/include/rcmail_output_html.php
@@ -1376,10 +1376,7 @@ EOF;
                 else if ($object == 'logo') {
                     $attrib += array('alt' => $this->xml_command(array('', 'object', 'name="productname"')));

-                    if (!empty($attrib['type']) && ($template_logo = $this->get_template_logo($attrib['type']
)) !== null) {
-                        $attrib['src'] = $template_logo;
-                    }
-                    else if (($template_logo = $this->get_template_logo()) !== null) {
+                    if (($template_logo = $this->get_template_logo($attrib['type'], empty($attrib['src']))) !== null) {
                         $attrib['src'] = $template_logo;
                     }

@@ -2488,10 +2485,11 @@ EOF;
      *
      * @param string  $type     Type of the logo to check for (e.g. 'print' or 'small')
      *                          default is null (no special type)
+     * @param boolean $strict   Use strict matching when looking for logos by type
      *
      * @return string image URL
      */
-    protected function get_template_logo($type = null)
+    protected function get_template_logo($type = null, $strict = true)
     {
         $template_logo = null;

@@ -2514,13 +2512,10 @@ EOF;
                     '*',
                 );

-                if (!empty($type)) {
-                    // Use strict matching, remove wild card options
-                    $template_names = preg_grep("/\*$/", $template_names, PREG_GREP_INVERT);
-                }
-                else {
-                    // No type set so remove those options from the list
-                    $template_names = preg_grep("/\\[\]$/", $template_names, PREG_GREP_INVERT);
+                // If a logo type is provided then limit the possibilities to that type
+                // otherwise remove the options with types from the list
+                if ($strict) {
+                    $template_names = preg_grep("/\]$/", $template_names, empty($type) ? PREG_GREP_INVERT : null);
                 }

                 foreach ($template_names as $key) {
```

**Comment by gianlucagiacometti** [comment]
@johndoh 

I'm extending my plugin roundcube-toolbox to add the functionality to customise the skin_logo parameter also per domain.
I will try those changes next week and let you know

**Comment by johndoh** [comment]
a workaround for this issue is to set [favicon] and [skin] to false. eg:
```
$config['skin_logo'] = array(
  'login[favicon]' => false,
  'login[small]' => false,
  'login' = '[a valid image]'
);
```

**Comment by gianlucagiacometti** [comment]
Thank you @johndoh for the suggestions.
I'm going to wait until you both decide about the new improvements for this. Since the issue is only in the login page, I choosed a login logo producing an unwanted favicon so confusing that it may seem done on purpose :-).
In the meanwhile I published my first version of the customisation tool for logos in my plugin roundcube-toolbox (BTW thank you, I learnt many things from you), which may provide a fast way for testing.
